src/System.Console/src/System/Console.cs

```diff
@@ -73,6 +90,41 @@ private static TextWriter CreateOutputWriter(Stream outputStream)
                     leaveOpen: true) { AutoFlush = true });
         }
 
+        private static bool _stdInRedirectQueried = false;
+        private static bool _stdOutRedirectQueried = false;
+        private static bool _stdErrRedirectQueried = false;
+
+        private static bool _isStdInRedirected;
+        private static bool _isStdOutRedirected;
+        private static bool _isStdErrRedirected;
+
+        public static bool IsInputRedirected
+        {
+            get
+            {
+                return Volatile.Read(ref _stdInRedirectQueried) ? _isStdInRedirected
+                    : EnsureInitialized(ref _isStdInRedirected, ref _stdInRedirectQueried, () => { return ConsolePal.IsInputRedirectedCore(); });
```

Comment:
> I am not sure how StrongBox is implemented but won't we pay unboxing penalty with it every time?

StrongBox is just a simple class like:
```C#
public class StrongBox<T>
{
    public T Value;
    public StrongBox() { }
    public StrongBox(T value) { Value = value; }
}
```

> Is not it?

Sorry, you're right, I was missing the difference in naming between _stdInRedirectQueried and _stdInRedirected.  Reading it quickly, I read them as the same value.

However, if you do stick with what you've done, you have a threading bug in your implementation.  You're setting _stdInRedirectQueried to true before you're setting _stdInRedirected; another thread could come along and see _stdInRedirectQueried as true and then proceed to read _stdInRedirected before you've actually initialized it.  That's one of the benefits of the ```StrongBox<T>``` approach: it's a single object, making it more difficult to get the threading matters wrong.  Normally I'd be against an approach that involved an extra allocation, but there's just a max of three of them for the whole process, so it's not a big deal, and it makes the code simpler IMHO, eliminated several fields, eliminating the extra initialization function, etc.